rfcs/syntax-static-imports.md

```diff
@@ -0,0 +1,281 @@
+!import "foo/bar/baz"
+```
+
+The `!import` used here is designed to mirror the current `export` used for exporting type annotations. To avoid ambiguity with function call syntax, the exclamation point is added, which lines up in style with the current `--!strict` style. This syntax may perhaps be extended in the future to cover other types of static declaration - perhaps even the aforementioned type checking mode.
+
+Unlike `require()` it is intended to introduce members to the namespace statically, not to return a value dynamically. This is why it is specifically a statement, not an expression:
+
+```Lua
+local foo = doSomething(require("foo")) -- ok
+local bar = doSomething(!import "foo") -- not ok
+```
+
+Since this statement is statically evaluated, the argument *must* be statically evaluatable (as is currently done to provide typechecking for `require()`):
+
+```Lua
+!import "foo" -- ok
+
+local bar = "foo"
+!import bar -- ok
+
+local bar = tostring(os.clock())
+!import bar -- not ok
+```
+
+The reason this is not specifically limited to string literals is to allow Roblox-like environments to evaluate statements for their imports, including the use of previously defined constant-like values:
+
+```Lua
+!import script.Parent.Libraries.Fusion
+
+local Package = script.Parent
+!import Package.Libraries.Fusion
+!import Package:FindFirstChild("Libraries"):FindFirstChild("Fusion")
+!import Package:WaitForChild("Libraries"):WaitForChild("Fusion")
+```
+
+### Syntax: Basic form
+
+The most basic form gives only the expression which resolves statically to the module path:
+
+```Lua
+!import "foo/bar/baz"
+```
+
+The returned value from the module is placed in a local variable, adopting the module's name. Types are placed in a namespace of the same name. This is how `require()` works today.
+
+The above snippet can be seen as equivalent to:
+
+```Lua
+local baz = require("foo/bar/baz")
+```
+
+### Syntax: Renaming
+
+To use a different name for the local variable/type namespace, an equals sign is added at the start of the statement:
+
+```Lua
+!import not_baz = "foo/bar/baz"
+```
+
+Beyond user convenience, this allows code to deal with modules named identically, as can be done today with `require()`.
+
+The above snippet can be seen as equivalent to:
+
+```Lua
+local not_baz = require("foo/bar/baz")
+```
+
+### Syntax: Type prefix
+
+To import only the types from a module, and skip evaluating the module at runtime, the `type` keyword can be added.
+
+```Lua
+!import type "foo/bar/baz"
+!import not_baz = type "foo/bar/baz"
+```
+
+This allows type information to be fetched from a module without introducing cyclic dependencies at runtime.
+
+There is currently no exactly equivalent code snippet for this. The closest is ['type smuggling' as presented by Anaminus](https://twitter.com/Anaminus/status/1585287008661938180), which does not work with exported types:
+
+```Lua
+type baz = typeof(require("foo/bar/baz"))
+```
+
+### Syntax: Local prefix
```

Comment:
I totally disagree with this.

- Static analysis tools like selene will now get significantly more complicated because now they need a total mapping of everything that can be imported. In the Roblox world this means that simple tools now need a complete sourcemap generated and provided. 
- Usual concerns of implicit globals in that it's not obvious at all where things are coming from as a reader.

Reply:
Implicit importing is completely opt-in from the top-down: you can tell from a code file that it will be importing stuff from a file, and you can see what pattern it will include (everything, or specific members). This information can then be used to find the members in the target module.

It is worth noting that we already have implicit imports. It happens when you `require()` a module that exports types. If your tools can't handle them, they already don't idiomatically handle Luau and I would consider this to be a tooling-level problem, not a problem with the spec (though I agree that effects on tooling should be considered as part of the RFC process, I do not think this is blocking)

Reply:
I have added this to the RFC as a drawback and discussion point.